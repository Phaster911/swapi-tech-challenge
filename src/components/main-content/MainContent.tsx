import styled from "styled-components";
import { COLORS, QUERIES } from "../../style-constants";
import MaxWidthWrapper from "../shared/max-width-wrapper/MaxWidthWrapper";
import MessageContainer from "../message-container/MessageContainer";
import CharacterTableSection from "../character-table-section/CharacterTableSection";
import Pagination from "../pagination/Pagination";
import { Person } from "../../react-query/queries/use-get-people.hook";

const StyledMain = styled.main`
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: auto;
  background-color: ${COLORS.mainBackground};
`;

const Wrapper = styled(MaxWidthWrapper)`
  margin-left: auto;
  margin-right: auto;
  padding: 1rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 1rem;
  height: 100%;

  @media ${QUERIES.tabletAndUp} {
    padding-top: 2rem;
    padding-bottom: 2rem;
    gap: 2rem;
  }
`;

export type MainContentProps = {
  shouldShowData: boolean;
  people: Person[];
  currentPage: number;
  pageCount: number;
  handleOnChangePage: (page: number) => void;
  isPending: boolean;
  isSuccess: boolean;
  isError: boolean;
  tryAgainHandler: () => void;
};

export default function MainContent({
  shouldShowData,
  people,
  currentPage,
  pageCount,
  handleOnChangePage,
  isPending,
  isSuccess,
  isError,
  tryAgainHandler,
}: MainContentProps) {
  return (
    <StyledMain>
      <Wrapper>
        {shouldShowData ? (
          <>
            <CharacterTableSection people={people} />
            <Pagination
              currentPage={currentPage}
              pageCount={pageCount}
              handleOnChangePage={handleOnChangePage}
            />
          </>
        ) : (
          <MessageContainer
            isPending={isPending}
            isSuccess={isSuccess}
            isError={isError}
            pageCount={pageCount}
            tryAgainHandler={tryAgainHandler}
          />
        )}
      </Wrapper>
    </StyledMain>
  );
}
