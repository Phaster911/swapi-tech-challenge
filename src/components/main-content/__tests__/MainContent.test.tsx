import { render, screen } from "@testing-library/react";
import MainContent, { MainContentProps } from "../MainContent";
import { Person } from "../../../react-query/queries/use-get-people.hook";

vi.mock("../../pagination/Pagination", () => ({
  default: () => <div>Pagination</div>,
}));

vi.mock("../../character-table-section/CharacterTableSection", () => ({
  default: () => <div>CharacterTableSection</div>,
}));

vi.mock("../../message-container/MessageContainer", () => ({
  default: () => <div>MessageContainer</div>,
}));

const person: Person = {
  id: "1",
  name: "name-1",
  gender: "gender-1",
  birthYear: "year-1",
  mass: "mass-1",
  filmsEntered: 0,
  height: "height-1",
};

describe("<MainContent />", () => {
  let props: MainContentProps;
  let handleOnChangePageFn: ReturnType<typeof vi.fn>;
  let tryAgainHandlerFn: ReturnType<typeof vi.fn>;
  beforeEach(() => {
    handleOnChangePageFn = vi.fn();
    tryAgainHandlerFn = vi.fn();
    props = {
      shouldShowData: true,
      people: [person],
      currentPage: 1,
      pageCount: 1,
      handleOnChangePage: handleOnChangePageFn,
      isPending: false,
      isSuccess: false,
      isError: false,
      tryAgainHandler: tryAgainHandlerFn,
    };
  });
  it("renders the component", () => {
    const { container } = render(<MainContent {...props} />);
    expect(container).toMatchSnapshot();
    expect(screen.getByText("CharacterTableSection")).toBeInTheDocument();
    expect(screen.getByText("Pagination")).toBeInTheDocument();
  });

  describe("when shouldShowData is false", () => {
    it("should render the message container", () => {
      const { container } = render(
        <MainContent {...props} shouldShowData={false} />
      );
      expect(container).toMatchSnapshot();
      expect(screen.getByText("MessageContainer")).toBeInTheDocument();
    });
  });
});
