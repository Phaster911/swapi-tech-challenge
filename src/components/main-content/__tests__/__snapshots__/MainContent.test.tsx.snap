// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<MainContent /> > renders the component 1`] = `
.c1 {
  width: clamp(16rem,95vw,90rem);
}

.c0 {
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: auto;
  background-color: #e9ecee;
}

.c2 {
  margin-left: auto;
  margin-right: auto;
  padding: 1rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 1rem;
  height: 100%;
}

@media (min-width: 750px) {
  .c2 {
    padding-top: 2rem;
    padding-bottom: 2rem;
    gap: 2rem;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div>
        CharacterTableSection
      </div>
      <div>
        Pagination
      </div>
    </div>
  </main>
</div>
`;

exports[`<MainContent /> > when shouldShowData is false > should render the message container 1`] = `
.c1 {
  width: clamp(16rem,95vw,90rem);
}

.c0 {
  flex-grow: 1;
  flex-shrink: 1;
  flex-basis: auto;
  background-color: #e9ecee;
}

.c2 {
  margin-left: auto;
  margin-right: auto;
  padding: 1rem;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  gap: 1rem;
  height: 100%;
}

@media (min-width: 750px) {
  .c2 {
    padding-top: 2rem;
    padding-bottom: 2rem;
    gap: 2rem;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1 c2"
    >
      <div>
        MessageContainer
      </div>
    </div>
  </main>
</div>
`;
